### ISSUE TITLE
Definitions from buildSrc/ not found in settings.gradle.kts using gradle 6.0-rc-1

### ISSUE THREAD

**Issue by wolfseifert** [gradle/gradle#11090]
Definitions from buildSrc/ not found in settings.gradle.kts using gradle 6.0-rc-1.
This still worked with gradle-6.0-20190827220029+0000 and earlier versions.

**Comment by ldaley** [comment]
Hi @wolfseifert, 

This has changed in 6.0, and was deprecated in 5.6. Please see: https://docs.gradle.org/current/userguide/upgrading_version_5.html#buildsrc_usage_in_gradle_settings

Can you please share some information about what kind of logic in buildSrc you are trying to use in the settings script?

**Comment by wolfseifert** [comment]
Hello  Luke,

 

here are some sample usages:

 

buildSrc/:

 



fun RepositoryHandler.nexus(): MavenArtifactRepository =
  maven {
    url = URI("http://nexushost:8081/repository/maven-public")
  }

fun Project.repositories() {
  repositories { nexus() }
}

fun PluginManagementSpec.repositories() {
  repositories { nexus() }
}

 
val osName: String =
  System.getProperty("os.name")!!

fun unsupportedOS(): Nothing =
  throw GradleException("""Operating system "$osName" is not supported.""")

enum class OS {
  LINUX, MACOSX, WINDOWS
}

val os: OS = when {
  osName.contains("linux", true) -> OS.LINUX
  osName.contains("mac", true) -> OS.MACOSX
  osName.contains("windows", true) -> OS.WINDOWS
  else -> unsupportedOS()
}

 

settings.gradle.kts:


 


pluginManagement {
  repositories()
}

 


if (os == OS.LINUX) include("gnome")


 

 

My intention was to configure this once (in buildSrc/) and then reuse it in different build.gradle.kts files.

The settings.gradle.kts is (in this sense) just another gradle build file.

 

Regards,

 

Wolfgang Seifert

 

 

 


Gesendet: Montag, 21. Oktober 2019 um 13:31 Uhr
Von: "Luke Daley" <notifications@github.com>
An: gradle/gradle <gradle@noreply.github.com>
Cc: wolfseifert <wolf.seifert@web.de>, Mention <mention@noreply.github.com>
Betreff: Re: [gradle/gradle] Definitions from buildSrc/ not found in settings.gradle.kts using gradle 6.0-rc-1 (#11090)


Hi @wolfseifert,

This has changed in 6.0, and was deprecated in 5.6. Please see: https://docs.gradle.org/current/userguide/upgrading_version_5.html#buildsrc_usage_in_gradle_settings

Can you please share some information about what kind of logic in buildSrc you are trying to use in the settings script?

**Comment by rschmitt** [comment]
I have a use case involving a large composite build, where code in `buildSrc` defines a list of packages that may be included in the build, `settings.gradle.kts` detects which of these packages are actually present and adds them to the build, and `build.gradle.kts` defines tasks for all included packages. What am I supposed to do on Gradle 6.0?

**Comment by ldaley** [comment]
Thanks for the info @wolfseifert. 

Unfortunately, there are pros and cons to both arrangements (`settings-then-buildSrc` and `buildSrc-then-settings`), and we opted for the former after considering.

The con of this is exactly what you have hit. It's now less convenient to use complex logic in your settings script. Now, you have to either:

1. Inline the logic into the settings file
2. Move the logic to a shared script that can be used where it needs to
3. Move the logic to a pre-built binary that you load in the settings file (i.e. a settings plugin)

The pros that compelled us to make the change:

1. Settings plugins can influence buildSrc and main build (i.e. apply a build plugin to both)
2. Build cache configuration is applied to buildSrc
3. buildSrc behaves more like a regular included build

We won't be changing the behaviour back to the pre Gradle 6 arrangement. Please let us know if you would like more detail on how to use one of the alternative mechanisms for using complex logic in a settings script.

**Comment by JavierSegoviaCordoba** [comment]
@ldaley do you have some examples of points 2 and 3?

I was declaring all the modules in buildSrc so now I have to declare two times, in settings and in buildSrc, can I avoid this?

**Comment by ldaley** [comment]
@JavierSegoviaCordoba can you please detail what you mean by “declaring all modules”. Exactly what that is will dictate the solution.

**Comment by JavierSegoviaCordoba** [comment]
I had a more elaborated version of val modules = listOf(app, core, ...) in buildSrc. So I can easily implement them in every build.gradle.kts which need them.

To easily include all in my project, in settings.gradle.kts I was doing a modules.forEach { module -> include(module) }

Now I have to declare them in the settings Gradle and in the buildSrc Kotlin object files.

**Comment by ldaley** [comment]
> So I can easily implement them in every build.gradle.kts which need them.

**Comment by JavierSegoviaCordoba** [comment]
If a module depends of other, for example I need the core module, I can put inside of dependencies: implementation(project(Modules.core))

**Comment by ldaley** [comment]
@JavierSegoviaCordoba I suggest moving your list of modules to a data file, and reading it in both places. For example, adding a `modules.properties` to your project root and enumerating it one way in settings and one in buildSrc.

**Comment by malachid** [comment]
Probably an important change to be calling out since it breaks anyone using resolutionStrategy with buildSrc for the versions.

**Comment by ldaley** [comment]
@malachid can you elaborate please.

**Comment by malachid** [comment]
Sure. In previous versions, we would use something like Versions.kt inside buildSrc to ensure a consistent version across dependencies, plugins, etc. We could have a resolutionStrategy that relied on that. Now, we have to have two copies of the versions - one inside settings.gradle.kts for any plugin that doesn't auto-resolve, and one in buildSrc to use for classpath/plugins/dependencies.

**Comment by ldaley** [comment]
@malachid what did your Versions.kt look like? Can you provide an indicative example?

**Comment by malachid** [comment]
If you look around you'll see that the most common recommendation for buildSrc over the last year was to migrate your Kotlin version information over rather than using strings everywhere.

So, for example, let's say you are writing a multi-module project. You might define `const val junitVersion` to point to Jupiter, then have an object or setOf for your test libs. In your various build.gradle.kts, you might do something like `TestLibs.forEach { testImplementation(it) }`

Now let's say a few of those modules are Kotlin multiplatform and rely on https://github.com/yshrsmz/BuildKonfig

When you apply plugin, it can't be found; so you add a resolutionStrategy to settings.gradle.kts that simply calls useModule for any that need it (using mapOf in the versions)...

say, something like this:

```
    resolutionStrategy {
        eachPlugin {
            Plugins.resolve
                    .filterKeys { it == requested.id.id }
                    .values
                    .forEach { useModule(it) }
        }
    }
```

With the change, you have to manually hard code the name and version of that plugin in settings.gradle.kts _as well as_ in your versions logic for the rest of the build; whereas before you only had one copy.

**Comment by ldaley** [comment]
Thanks @malachid. Your additional information is useful.

**Comment by bejibx** [comment]
I have pretty much the same configuration as @malachid in my work project. I spent a lot of time in past to refactor it to this architecture and it works pretty good for us. It is very convenient to have all your dependencies and versions in one place. This makes updating them super easy - you just changes version there and that's all. Also we are using Gradle Kotlin DSL and declare our dependencies/version as properties inside Kotlin object, which also have many benefits.

Now I'm trying to update Gradle to latest version and suddenly my whole architecture is gone. Now I have to redeclare some dependencies (plugins, repositories...) in settings.gradle.kts. Now every time you want to update some plugin, you have to remember to change it version in two places.

Does Gradle 6.+ have other mechanism to achieve similar thing?

**Comment by nkiesel** [comment]
same here: I use the "buildSrcverions" plugin which generates a `buildSrc/src/main/kotlin/Versions.kt` file containing `const val org_jlleitschuh_ktlint_gradle_plugin: String = "9.1.1"`. I currently then use that in `settings.gradle` like so: `id "org.jlleitschuh.gradle.ktlint" version Versions.org_jlleitschuh_ktlint_gradle_plugin`. This no longer works.  Iguess I have to look at newer version of "buildSrcVersions" which might remedy this, but it will for sure mean that I have to re-tran my team to understand the new version mgmt.

**Comment by nanjingdaqi** [comment]
> Thanks for the info @wolfseifert.
> 
> Unfortunately, there are pros and cons to both arrangements (`settings-then-buildSrc` and `buildSrc-then-settings`), and we opted for the former after considering.
> 
> The con of this is exactly what you have hit. It's now less convenient to use complex logic in your settings script. Now, you have to either:
> 
> 1. Inline the logic into the settings file
> 2. Move the logic to a shared script that can be used where it needs to
> 3. Move the logic to a pre-built binary that you load in the settings file (i.e. a settings plugin)
> 
> The pros that compelled us to make the change:
> 
> 1. Settings plugins can influence buildSrc and main build (i.e. apply a build plugin to both)
> 2. Build cache configuration is applied to buildSrc
> 3. buildSrc behaves more like a regular included build
> 
> We won't be changing the behaviour back to the pre Gradle 6 arrangement. Please let us know if you would like more detail on how to use one of the alternative mechanisms for using complex logic in a settings script.

I am using gradle kts. Could anyone provide the demo for "shared project" and "setting plugin"? I failed to search relevant information.

**Comment by dybarsky** [comment]
Hi here. 
We also faced this problem. In our multimodule project we have different kotlin files in `buildSrc` dir. For example:
```
object AppModule {
  const val APPLICATION = ":application"
  const val BASE_API = ":base-api"
  const val BASE_EXTENSION = ":base-extension"
  const val BASE_GEOLOCATION = ":base-geolocation"
  ...
}

object AppLibrary {
    const val OKHTTP = "com.squareup.okhttp3:okhttp:${AppVersion.OKHTTP}"
}

object AppVersion {
  const val OKHTTP = "4.4.1"
}
```

Our `settings.gradle` file in root project uses `AppModule` class like this
```
include(
  AppModule.APPLICATION,
  AppModule.BASE_API,
  AppModule.BASE_EXTENSIONS,
   ...
)
```

With new gradle version we can't use `AppModule` class anymore and should use module names as strings. Any ideas how we can solve this without duplication of modules definitions or reading from properties file?

**Comment by JavierSegoviaCordoba** [comment]
FileWalkTree to find all the modules or use the auto-module plugin

**Comment by eygraber** [comment]
@ldaley any suggestions for how to handle the issue that @malachid presented? I run into this issue a lot (and it appears that others do as well).

**Comment by android10** [comment]
Same case here. In order to have dependencies in one place defined inside ```buildSrc```.
For examples ```Dependencies.kt``` where we define modules:

```kotlin
object Modules {
    const val app = ":app"
}
```

From ```settings.gradle.kts```:

```kotlin
include (Modules.app)  -> ERROR: Unresolved reference: Modules
//include (":app")     -> Works but I do not want string literals here
```

I'm not expecting another API change if the decision has been made but at least some **tips/tricks/pattern** on how to handle this scenario. 

Thank you very much :heart:

**Comment by JavierSegoviaCordoba** [comment]
@android10 take a look to [auto-module](https://github.com/pablisco/auto-module).

Or you can create an algorithm to `include` every module automatically using [`FileWalkTree`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/-file-tree-walk/) and searching every `build.gradle.kts` file in the project (excluding the root `build.gradle.kts` file).

**Comment by dybarsky** [comment]
@JavierSegoviaCordoba I tried both options. It's terrible over-complicated solution.

**Comment by JavierSegoviaCordoba** [comment]
@dybarsky `auto-module` option is a one-line solution that even creates for you the `Modules.kt` with all the modules of the project automatically. It is easier than doing manually, pre-Gradle 6 even.

**Comment by dybarsky** [comment]
@JavierSegoviaCordoba It's onle line solution, it's true. But it adds task, generates/modifies gradle files, there is dependency handling. That's not what we need.

**Comment by dybarsky** [comment]
```kotlin
rootDir.walk().maxDepth(1)
  .filter {
    it.isDirectory && file("${it.absolutePath}/build.gradle.kts").exists()
  }
  .forEach {
    include(":${it.name}")
  }
```
@android10 That's solution which works for us. 
@JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class

**Comment by dybarsky** [comment]
I faced new issue. We extracted plugins version to `buildSrc` in same way as `Modules.kt`. Not I have problems with applying logic to plugins resolution strategy in `settings.gradle` file.
```kotlin
pluginManagement {
  repositories {
    google()
    mavenCentral()
    gradlePluginPortal()
  }
  resolutionStrategy {
    eachPlugin {
      // https://docs.gradle.org/current/userguide/kotlin_dsl.html#sec:plugins_resolution_strategy
      if (requested.id.id.startsWith("com.android.")) {
        useModule(BuildLibrary.androidGradle)
      }
      if (requested.id.id.startsWith("kotlin")) {
        useModule(BuildLibrary.kotlinGradle)
      }
      if (requested.id.id.startsWith("com.google.firebase.firebase-perf")) {
        useModule(BuildLibrary.firebasePerformance)
      }
    }
  }
}
```
When apply `useModule` build failed, because can't find `BuildLibrary` defined in `buildSrc`
```kotlin
object BuildLibrary {
  const val androidGradle = "com.android.tools.build:gradle:${BuildVersion.ANDROID_GRADLE}"
}
```
Any ideas how to solve it? Thanks.

**Comment by android10** [comment]
@dybarsky @JavierSegoviaCordoba thanks both of you for the contribution :heart:

**Comment by Dreyar** [comment]
> ```kotlin
> rootDir
>   .walk()
>   .maxDepth(1)
>   .filter {
>     it.name != "buildSrc"
>       && it.isDirectory 
>       && file("${it.absolutePath}/build.gradle.kts").exists()
>   }
>   .forEach {
>     include(":${it.name}")
>   }
> ```
> 
> @android10 That's solution which works for us.
> @JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class

How did you used this to fix your issue above? (https://github.com/gradle/gradle/issues/11090#issuecomment-632019679)
I changed gradle from 6.5 to 5.6.4 but i do not like that approach.

**Comment by dybarsky** [comment]
@Dreyar I have replaced `include(AppModule.APPLICATION, ...)` with logic above. Now settings.gradle.kts file has no references to classes from buildSrc scope.

**Comment by sellmair** [comment]
I also would like to add that this is causing inconveniences for me in certain projects: 
E.g. in https://github.com/Kotlin/dokka, we would like to have some custom build logic regarding the repositories where we get our dependencies from. Now I have to duplicate this in settings.gradle.kts, because the same logic should be used in the `pluginManagement` block (which can't access our buildSrc folder) 😢

**Comment by Blacklands** [comment]
I've read all of this, and I'm still confused.

First I read that it's preferred to use the new `plugins {}` block to apply plugins.

Then I read that it's preferred to apply plugins in `settings.gradle[.kts]`, because the `plugins { }` doesn't allow (most) custom logic, by design. In the settings file, the `pluginManagement { }` block is there specifically to allow custom logic.

Up to this point, that makes sense to me. But if `settings.gradle[.kts]` is basically the very first thing that is being evaluated/built in a project, then where should that custom logic come from, if I don't want to put it all right into the file itself...?

[This comment](https://github.com/gradle/gradle/issues/11090#issuecomment-589931599) said:

```
2. Move the logic to a shared script that can be used where it needs to
3. Move the logic to a pre-built binary that you load in the settings file (i.e. a settings plugin)
```

Does 3 mean that one would need to create a plugin, compile it, and then include it in `settings.gradle[.kts]` somehow (and how?)? That sounds very overkill if one just has some simple build logic (that may also be project-specific, so the plugin couldn't even be reused somewhere else).

For 2, what is a "shared script"? I couldn't find anything about that, anywhere. Is it a Gradle script? If it is, wouldn't it have the same problem as `buildSrc`, being built only _after_ `settings.gradle[.kts]` is already done?

Overally, it feels like there is a bit of a gap here now. Maybe something else is needed, something that is specifically made to include custom logic into `settings.gradle[.kts]`, and is built before it? But I suppose that it's somewhat of a chicken-and-egg problem... something needs to be the _first_ thing to be built, after all, so that would just move the problem?

**Comment by malachid** [comment]
> Up to this point, that makes sense to me. But if `settings.gradle[.kts]` is basically the very first thing that is being evaluated/built in a project, then where should that custom logic come from, if I don't want to put it all right into the file itself...?
> ...
> Overally, it feels like there is a bit of a gap here now. Maybe something else is needed, something that is specifically made to include custom logic into `settings.gradle[.kts]`, and is built before it? But I suppose that it's somewhat of a chicken-and-egg problem... something needs to be the _first_ thing to be built, after all, so that would just move the problem?

I think that's the cruz. Pre-6, the _first_ was buildSrc.  6 made a breaking change to the order of operations so that it was no longer first.

I'm a bit resigned to the fact I will have to copy/pasta everything to do with plugins from the buildSrc into the settings.gradle.kts and maintain two copies.  This has, unfortunately, prevented us from migrating to 6 as it is a large tech debt (multiple repos to refactor) with extremely low priority.  We know we eventually have to do it - but it is hard to justify at this point.

**Comment by floscher** [comment]
This is how I worked around this limitation:

I create a file `${rootProject.projectDir}/buildSrc/buildSrc/build.gradle.kts`:
```kotlin
plugins {
  `kotlin-dsl`
}
repositories.jcenter()
sourceSets.main {
  java {
    setSrcDirs(setOf(projectDir.parentFile.resolve("src/main/kotlin")))
    include("VersionNumber.kt")
  }
}
```
This makes that the `VersionNumber` class is compiled one additional time before the `buildSrc` project is evaluated. I can then use that class in `${rootProject.projectDir}/buildSrc/build.gradle.kts`.

I can then use the plugin versions in my `${rootProject.projectDir}/buildSrc/build.gradle.kts` like this:
```kotlin
dependencies {
  implementation("plugin.group:plugin.module:${VersionNumber.PLUGIN}")
}
```

And in subprojects I can write the plugin definitions without version numbers:
```kotlin
plugins {
  id("plugin.id")
}
```
In fact Gradle will then even complain when I add `.version("1.2.3")`.

See also Kotlin/dokka#1321 for an example of this approach.

**Comment by sellmair** [comment]
😮 Impressive workaround!

**Comment by jonecx** [comment]
> ```kotlin
> rootDir
>   .walk()
>   .maxDepth(1)
>   .filter {
>     it.name != "buildSrc"
>       && it.isDirectory 
>       && file("${it.absolutePath}/build.gradle.kts").exists()
>   }
>   .forEach {
>     include(":${it.name}")
>   }
> ```
> 
> @android10 That's solution which works for us.
> @JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class

Do I need to put put this thing as is in the settings.gradle.kts?

**Comment by jonecx** [comment]
> Same case here. In order to have dependencies in one place defined inside `buildSrc`.
> For examples `Dependencies.kt` where we define modules:
> 
> ```kotlin
> object Modules {
>     const val app = ":app"
> }
> ```
> 
> From `settings.gradle.kts`:
> 
> ```kotlin
> include (Modules.app)  -> ERROR: Unresolved reference: Modules
> //include (":app")     -> Works but I do not want string literals here
> ```
> 
> I'm not expecting another API change if the decision has been made but at least some **tips/tricks/pattern** on how to handle this scenario.
> 
> Thank you very much ❤️

@android10 were you able get around this one 🤔?

**Comment by roths** [comment]
I am confuse.
This way can work:
`settings.gradle` code like this:
```
gradle.beforeProject { project ->
    if (project == project.rootProject) {
        project.with {
            apply from: "plugin.gradle"
        }
    }
}
```
`plugin.gradle` content like this:
```
project.with {
            buildscript.dependencies.with {
                classpath Deps.plugin
            }
}
```
Deps.kt file is in buildSrc dir.
But, when i put the logic from `plugin.gradle` into `gradle.beforeProject`
`settings.gradle` code like this:
```
gradle.beforeProject { project ->
    if (project == project.rootProject) {
        project.with {
            buildscript.dependencies.with {
                classpath Deps.plugin
            }
        }
    }
}
```
i will get an error: Could not get unknown property 'Deps'.
Could anyone explain it?

**Comment by floscher** [comment]
@roths You can't write `Deps.plugin` anywhere in your `settings.gradle`file, because the `Deps` class is compiled and put on the classpath only after the `settings.gradle` script is evaluated.

The reason it works when you put it in a separate file and apply it inside a `gradle.beforeProject{}` block is, that the content of the `plugin.gradle` script is only evaluated as soon as the `build.gradle` scripts are ready to be evaluated. At that point the classes from your `buildSrc/` directory are on the classpath and can be used in the `plugin.gradle` script.

But essentially I think your solution is equivalent to just writing
```groovy
buildscript.dependencies.with {
  classpath Deps.plugin
}
```
at the beginning of your root `build.gradle` file.

---

Above in https://github.com/gradle/gradle/issues/11090#issuecomment-674798423 I describe a similar way of managing plugin versions. My approach defines the plugin versions in `buildSrc/build.gradle.kts`.

<details>
<summary>For your project it would look something like this:

(click this text to expand :arrow_up_down: )</summary>

Add the file `${rootProject.projectDir}/buildSrc/buildSrc/build.gradle.kts`:
```
plugins {
  `kotlin-dsl`
}
repositories.jcenter()
sourceSets.main {
  java {
    setSrcDirs(setOf(projectDir.parentFile.resolve("src/main/kotlin")))
    include("Deps.kt")
  }
}
```
Add the following to `${rootProject.projectDir}/buildSrc/buildSrc/build.gradle.kts`:
```kotlin
dependencies {
  implementation(Deps.plugin)
}
```
Then you can add plugins without version number:
```kotlin
plugins {
  id("plugin.id")
}
```</details>

**Comment by kubapet** [comment]
I don't really like this change. I think the buildSrc should have remaind the sources that are compiled all the way before the project is even loaded not somewhere in the middle.

We have a project with around 80 subprojects and that depends on each other. Originally I had the definitions of the subproject just in the buildSrc, I cannot use them in settings.gradle.kts anymore and it must be "stringly" typed in an ugly way..

buildSrc/src/main/kotlin/.../Modules.kt:
```
object Modules {
  val framework_api by ModuleNameDelegate
  val framework_api_dto by ModuleNameDelegate
  val framework_backend by ModuleNameDelegate
  val framework_common by ModuleNameDelegate
  val framework_frontend by ModuleNameDelegate
  val framework_itest by ModuleNameDelegate
  val framework_test by ModuleNameDelegate
  val framework_testdb by ModuleNameDelegate
```

and settings.gradle.kts:
```
include(":framework-common")
include(":framework-api")
include(":framework-api-dto")
include(":framework-backend")
include(":framework-frontend")
include(":framework-itest")
include(":framework-test")
include(":framework-testdb")
```

**Comment by TurKurT656** [comment]
> ```kotlin
> rootDir
>   .walk()
>   .maxDepth(1)
>   .filter {
>     it.name != "buildSrc"
>       && it.isDirectory 
>       && file("${it.absolutePath}/build.gradle.kts").exists()
>   }
>   .forEach {
>     include(":${it.name}")
>   }
> ```
> 
> @android10 That's solution which works for us.
> @JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class

This cannot handle modules like ":library:core". It's just one depth level. How can fix this issue?

**Comment by JavierSegoviaCordoba** [comment]
> > ```kotlin
> > rootDir
> >   .walk()
> >   .maxDepth(1)
> >   .filter {
> >     it.name != "buildSrc"
> >       && it.isDirectory 
> >       && file("${it.absolutePath}/build.gradle.kts").exists()
> >   }
> >   .forEach {
> >     include(":${it.name}")
> >   }
> > ```
> > 
> > 
> > @android10 That's solution which works for us.
> > @JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class
> 
> This cannot handle modules like ":library:core". It's just one depth level. How can fix this issue?

Remove `maxDepth(1)`

**Comment by TurKurT656** [comment]
> > > ```kotlin
> > > rootDir
> > >   .walk()
> > >   .maxDepth(1)
> > >   .filter {
> > >     it.name != "buildSrc"
> > >       && it.isDirectory 
> > >       && file("${it.absolutePath}/build.gradle.kts").exists()
> > >   }
> > >   .forEach {
> > >     include(":${it.name}")
> > >   }
> > > ```
> > > 
> > > 
> > > @android10 That's solution which works for us.
> > > @JavierSegoviaCordoba Thanks for pointing at `FileWalkTree` class
> > 
> > 
> > This cannot handle modules like ":library:core". It's just one depth level. How can fix this issue?
> 
> Remove `maxDepth(1)`

It gives me this error:

> A problem occurred evaluating project ':app'.
> Project with path ':data:remote' could not be found in project ':app'.

**Comment by JavierSegoviaCordoba** [comment]
Just try automodule by pablisco

**Comment by jjohannes** [comment]
Since there are so many comments still on this issue, let me clarify a few things about where we are and where we are actively moving to right now.

For the general topic "I want buildSrc to be done before anything else" there will be a solution soon [with included builds and setting plugins](https://github.com/gradle/gradle/issues/14562). Most likely in the next release. Then you will be able to include a build, using a new DSL method, that is available earlier. This build can then contain a settings plugin which you can apply in settings. This makes the Jar containing the plugin available on the settings classpath. (Although ideally you would define a plugin extension and not use classes/static methods directly.)

*settings.gradle.kts*
``` 
pluginManagement {
    includeBuildEarly("my-build-logic") //<- WIP - new API under development
}
plugins {
   apply("my.settings.plugin) // this is defined in the "my-build-logic" build
}
```

For the topic of sharing dependencies and versions, we are also [working on general improvements](https://github.com/gradle/gradle/pull/14896) that might make some "custom solutions" unnecessary in the future.

Generally, you should try to avoid using `buildscript {}` or using  `resolutionStrategy {}` in settings. Instead you can define plugins in included builds. You can then manage all dependencies (also to plugins) in the build files of these builds.

If you want to share constants between all these builds, you can have one build only for these constants that you include in all others. Like [the `libraries` build in this example.](https://github.com/jjohannes/idiomatic-gradle/tree/master/libraries).

See also the sample about [sharing convention plugins with build logic build](https://docs.gradle.org/current/samples/sample_sharing_convention_plugins_with_build_logic.html).

Hope these are some helpful pointers.

**Comment by keith-miller** [comment]
Hello @jjohannes, does one of those examples show how to use a single source of truth for plugin versions when using convention plugins? The `idiomatic-gradle` doesn't seem to follow the convention plugins example in the Gradle docs, so I'm a bit confused: https://docs.gradle.org/current/samples/sample_building_java_applications_multi_project.html

Sorry, new to using Kotlin with Gradle and my Google search landed me here. Thanks!

**Comment by jjohannes** [comment]
@keith-miller the Java application sample does not contain anything on the topic of sharing plugin versions.

In the 6.8 docs we have a more extended example, which shows, among other things, how you can share versions via a separate build containing _platform_ project(s). It includes sharing the plugin versions.
https://docs.gradle.org/6.8-rc-1/samples/sample_structuring_software_projects.html

**Comment by keith-miller** [comment]
Thank you!

**Comment by nikospamp** [comment]
@jjohannes I don't know if I miss something, but the kotlin DSL that's located [here](https://docs.gradle.org/6.8-rc-1/samples/sample_structuring_software_projects.html) is not passing build. I have been tinkering with the gradles for quite some time now but no luck.

**Comment by jjohannes** [comment]
@nikospamp this looks like IDEA is not using the right Gradle version. Did you use the latest version of the sample? https://docs.gradle.org/current/samples/sample_structuring_software_projects.html

Did you open the root folder of the sample ("umbrella" build)? Only then IDEA picks up the version from the Gradle wrapper as it is located in the root in this example. Otherwise, you have to configure the Gradle version manually: `Preferences -> Build, ... -> Gradle -> Use Gradle from:`

**Comment by nikospamp** [comment]
@jjohannes I am using the projects gradle, gradle-6.8.3-bin and I opened the root folder, not an underline module. 

If you have any ideas please do tell, otherwise I will post here the solution if I come up with one.

**Comment by jjohannes** [comment]
@nikospamp the error looks a lot like IDEA is using the wrong Gradle version as `dependencyResolutionManagement {}`, which it can't find, is a rather new DSL element in settings. Have you double checked `Preferences -> Build, ... -> Gradle -> Use Gradle from:`?

Do you get the error when syncing already, or when you run a build?

If it is only when running a build, you may need to adjust the run configuration IDEA creates, so that it uses the build from the root. Otherwise, IDEA tries to find the wrapper inside the "child" build you are running a task from. (This is something that will hopefully improve in IDEA soon.)

Or you configure `Preferences -> Build, ... -> Gradle -> Use Gradle from:` to use a Gradle installation instead of the wrapper.

**Comment by trevor-hackman-kr** [comment]
> Thanks for the info @wolfseifert.
> 
> Unfortunately, there are pros and cons to both arrangements (`settings-then-buildSrc` and `buildSrc-then-settings`), and we opted for the former after considering.
> 
> The con of this is exactly what you have hit. It's now less convenient to use complex logic in your settings script. Now, you have to either:
> 
> 1. Inline the logic into the settings file
> 2. Move the logic to a shared script that can be used where it needs to
> 3. Move the logic to a pre-built binary that you load in the settings file (i.e. a settings plugin)
> 
> The pros that compelled us to make the change:
> 
> 1. Settings plugins can influence buildSrc and main build (i.e. apply a build plugin to both)
> 2. Build cache configuration is applied to buildSrc
> 3. buildSrc behaves more like a regular included build
> 
> We won't be changing the behaviour back to the pre Gradle 6 arrangement. Please let us know if you would like more detail on how to use one of the alternative mechanisms for using complex logic in a settings script.

This was an awful decision and makes settings.gradle awful to use.

**Comment by trevor-hackman-kr** [comment]
> Thanks for the info @wolfseifert.
> 
> Unfortunately, there are pros and cons to both arrangements (`settings-then-buildSrc` and `buildSrc-then-settings`), and we opted for the former after considering.
> 
> The con of this is exactly what you have hit. It's now less convenient to use complex logic in your settings script. Now, you have to either:
> 
> 1. Inline the logic into the settings file
> 2. Move the logic to a shared script that can be used where it needs to
> 3. Move the logic to a pre-built binary that you load in the settings file (i.e. a settings plugin)
> 
> The pros that compelled us to make the change:
> 
> 1. Settings plugins can influence buildSrc and main build (i.e. apply a build plugin to both)
> 2. Build cache configuration is applied to buildSrc
> 3. buildSrc behaves more like a regular included build
> 
> We won't be changing the behaviour back to the pre Gradle 6 arrangement. Please let us know if you would like more detail on how to use one of the alternative mechanisms for using complex logic in a settings script.

This was an awful decision.

**Comment by JacquesSmuts** [comment]
Something you can do if it fits your circumstance:

`gradle.properties`
```
kotlin.version=1.8.20
```

`buildSrc/src/main/kotlin/Versions.kt`
```
object Versions {
    object Kotlin {
        val version = System.getProperty("kotlin.version")
    }
}
```

**Comment by lehvolk** [comment]
This was an awful decision. I have custom repositories for CI and local env. Have idea to have a common function to detect that build is on CI and use global repos for local env. Now I have to copy-paste code in 3 places instead of one common place.
